### PR TITLE
Custom UNet architecture

### PR DESCRIPTION
**Pull request by lkdci**
Framework to create UNet like architectures with large degree of freedom and architecture params.

### CODE REVIEW

File: src/super_gradients/training/models/model_factory.py [hunk 1]
```diff
@@ -20,6 +23,25 @@
 logger = get_logger(__name__)
 
 
+def get_arch_params(config_name, overriding_params: Dict = None) -> Dict:
```
Comment (lkdci) [inline src/super_gradients/training/models/model_factory.py:26]: I'll really appreciate your review here, @shaydeci

File: src/super_gradients/training/models/segmentation_models/unet/unet.py [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+    def _separate_lr_multiply_params(self):
+        multiply_lr_params, no_multiply_params = {}, {}
+        for name, param in self.named_parameters():
+            if "backbone." in name:
+                no_multiply_params[name] = param
+            else:
+                multiply_lr_params[name] = param
+        return multiply_lr_params.items(), no_multiply_params.items()
+
+    def prep_model_for_conversion(self,
```
Comment (ofrimasad) [inline src/super_gradients/training/models/segmentation_models/unet/unet.py:203]: `prep_model_for_coversion` has a defined signature that allows the convertor to convert it.
`    def prep_model_for_conversion(self, input_size: Union[tuple, list] = None, **kwargs):`
you can keep the append_sigmoid and append_softmax, but please add input_size

File: src/super_gradients/training/models/segmentation_models/unet/unet_decoder.py [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+        raise NotImplementedError(f"block_id: {block_id} is not a valid option.")
+
+
+class Decoder(nn.Module):
+    def __init__(self,
+                 skip_channels_list: List[int],
+                 up_block_repeat_list: List[int],
+                 skip_expansion: float,
+                 decoder_scale: float,
+                 up_block_types: List[Union[UpBlockType, int]],
```
Comment (ofrimasad) [inline src/super_gradients/training/models/segmentation_models/unet/unet_decoder.py:105]: why do you allow int here? (if because of NAS then remove this)

File: src/super_gradients/training/models/segmentation_models/unet/unet_decoder.py [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+            nn.Sequential(*[ConvBNReLU(out_channels, out_channels, kernel_size=3, padding=1, bias=False) for _ in range(num_repeats - 1)]),
+        )
+
+    def forward(self, x, skip):
+        x = self.up_path(x)
+        x = torch.cat([x, skip], dim=1)
+        return self.last_convs(x)
+
+
+class UpBlockType(Enum):
```
Comment (ofrimasad) [inline src/super_gradients/training/models/segmentation_models/unet/unet_decoder.py:80]: please follow the logic of factories:
- define a dictionary of possible UpFuseBlocks (possibly, move them all to another file)
- define a factory 
- use a decorator over Decoder.__init__() to create a UpFuseBlock from a string

File: src/super_gradients/training/models/segmentation_models/unet/unet_decoder.py [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+        self.up_stages = nn.ModuleList()
+        in_channels = skip_channels_list.pop(0)
+        skip_channels_list.append(None)
+        for i in range(len(up_block_types)):
+            self.up_stages.append(
+                up_block_types[i].up_block_cls(in_channels, skip_channels_list[i], self.up_channels_list[i],
+                                               num_repeats=up_block_repeat_list[i], **up_block_kwargs))
+            in_channels = self.up_channels_list[i]
+
+    def make_skip_projection(self,
```
Comment (ofrimasad) [inline src/super_gradients/training/models/segmentation_models/unet/unet_decoder.py:154]: please make it private (_)

File: src/super_gradients/training/models/segmentation_models/unet/unet_encoder.py [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+        if group_width > inter_channels:
+            return inter_channels
+        group_pow = int(math.log2(group_width))
+        for pow in range(group_pow, -1, -1):
+            if (inter_channels / 2 ** pow) % 1 == 0:
+                return int(2 ** pow)
+        return 1
+
+
+class BlockType(Enum):
```
Comment (ofrimasad) [inline src/super_gradients/training/models/segmentation_models/unet/unet_encoder.py:159]: here too - use factories instead of enums

**Review by ofrimasad — COMMENTED** [review]
looks ok but we need to make more according to SG logic.
my comment inside

File: src/super_gradients/recipes/arch_params/unet_default_arch_params.yaml [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+head_hidden_channels:
+
+use_aux_heads: False
+aux_heads_params:
+  use_aux_list: [False, False, True, True, True]
+  aux_heads_factor: [2, 4, 8, 16, 32]
+  aux_hidden_channels: [32, 32, 64, 64, 64]
+  aux_out_channels: [1, 1, 19, 19, 19]
+
+_convert_: all
```
[the remark this answers is not in the document]
Comment (lkdci) [reply src/super_gradients/recipes/arch_params/unet_default_arch_params.yaml:51]: Done [30f13c6](https://github.com/Deci-AI/super-gradients/pull/401/commits/30f13c608c9be02306e2547bf66020dc4d5f8e91)

File: src/super_gradients/training/models/model_factory.py [hunk 8]
```diff
@@ -20,6 +23,25 @@
...
+    """
+    Class for creating arch parameters dictionary, taking defaults from yaml
```
[the remark this answers is not in the document]
Comment (lkdci) [reply src/super_gradients/training/models/model_factory.py]: Moved to arch_params_factory [ae186b5](https://github.com/Deci-AI/super-gradients/pull/401/commits/ae186b536058b58b9aa2dccac322c4b6c0c7e252)

File: src/super_gradients/training/models/segmentation_models/context_modules.py [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+from super_gradients.training.utils.module_utils import ConvBNReLU
+from abc import ABC, abstractmethod
+
+
+class AbstractContextModule(ABC):
+    @abstractmethod
+    def output_channels(self):
+        raise NotImplementedError
+
+
```
[the remark this answers is not in the document]
Comment (lkdci) [reply src/super_gradients/training/models/segmentation_models/context_modules.py:16]: Done, moved sppm to context_module. [d24d333](https://github.com/Deci-AI/super-gradients/pull/401/commits/d24d3333d958ee47af96ea947f17cbad6bfd0d36)

File: src/super_gradients/training/models/segmentation_models/context_modules.py [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+class ASPP(nn.Module, AbstractContextModule):
```
[the remark this answers is not in the document]
Comment (lkdci) [reply src/super_gradients/training/models/segmentation_models/context_modules.py]: Done. [d24d333](https://github.com/Deci-AI/super-gradients/pull/401/commits/d24d3333d958ee47af96ea947f17cbad6bfd0d36)

File: src/super_gradients/training/models/segmentation_models/unet/unet.py [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+class UNetBase(SegmentationModule):
+    def __init__(self,
+                 num_classes: int,
+                 use_aux_heads: bool,
+                 final_upsample_factor: int,
+                 head_hidden_channels: Optional[int],
+                 head_upsample_mode: Union[UpsampleMode, str],
+                 align_corners: bool,
+                 backbone_params: dict,
+                 context_module_name: Optional[str],
```
[the remark this answers is not in the document]
Comment (lkdci) [reply src/super_gradients/training/models/segmentation_models/unet/unet.py]: A little bit problematic since the `in_channels` arg is dynamic and depends on the backbone output.
I took OptimizerTypeFactory as example. Here [83ef910](https://github.com/Deci-AI/super-gradients/pull/401/commits/83ef910b7f673939f45d9bee095f0c9400228b1b)

File: src/super_gradients/training/models/segmentation_models/unet/unet.py (hunk 2 above)
Comment (lkdci) [reply src/super_gradients/training/models/segmentation_models/unet/unet.py:204]: Done [1cf7d2b](https://github.com/Deci-AI/super-gradients/pull/401/commits/1cf7d2bbf5ea0d9ba0172513d78f1e24655757e1)

File: src/super_gradients/training/models/segmentation_models/context_modules.py [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+        x = torch.cat([dilated_conv(x) for dilated_conv in self.dilated_conv_list], dim=1)
+        return x
+
+
+CONTEXT_TYPE_DICT: MutableMapping[str, Type[AbstractContextModule]] = {
+    "aspp": ASPP,
+}
+
+
+def build_context_module(context_module_name: str, context_module_params: dict, in_channels: int):
```
[the remark this answers is not in the document]
Comment (lkdci) [reply src/super_gradients/training/models/segmentation_models/context_modules.py]: A little bit problematic since the in_channels arg is dynamic and depends on the backbone output.
I took OptimizerTypeFactory as example. Here [83ef910](https://github.com/Deci-AI/super-gradients/pull/401/commits/83ef910b7f673939f45d9bee095f0c9400228b1b)

File: src/super_gradients/training/models/segmentation_models/unet/unet_decoder.py (hunk 4 above)
Comment (lkdci) [reply src/super_gradients/training/models/segmentation_models/unet/unet_decoder.py:82]: Since the params for the blocks are dynamic parameters that depends on other layers in the models, such as in_channels, the factory is not required to initiate an object of the block class, only required to return a class type from string. For this matter I created a factory class called `TypeFactory` which can be initiated from enum or from a dictionary. [f33631f](https://github.com/Deci-AI/super-gradients/pull/401/commits/f33631f226d77345abd9db0c38b369015efc6e17)

File: src/super_gradients/training/models/segmentation_models/unet/unet_encoder.py (hunk 6 above)
Comment (lkdci) [reply src/super_gradients/training/models/segmentation_models/unet/unet_encoder.py]: Done

File: src/super_gradients/training/models/segmentation_models/unet/unet_decoder.py (hunk 5 above)
Comment (lkdci) [reply src/super_gradients/training/models/segmentation_models/unet/unet_decoder.py]: Done [0a56285](https://github.com/Deci-AI/super-gradients/pull/401/commits/0a56285e7b3123ad1b0c377e28f4589398942969)

File: src/super_gradients/training/models/segmentation_models/unet/unet_decoder.py (hunk 3 above)
Comment (lkdci) [reply src/super_gradients/training/models/segmentation_models/unet/unet_decoder.py]: Removed [f33631f](https://github.com/Deci-AI/super-gradients/pull/401/commits/f33631f226d77345abd9db0c38b369015efc6e17)

**Comment by lkdci** [comment]
Added changes according to the review, see implementations of factories @ofrimasad

File: src/super_gradients/recipes/arch_params/unet_default_arch_params.yaml [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+    anti_alias: True                # RepVGG stage param
+    droppath_prob: 0.               # XBlock stage param
+    bottleneck_ratio: 1.            # XBlock stage param
+    group_width: 16                 # XBlock stage param
+    se_ratio:                       # XBlock stage param
+    steps: 4                        # STDC stage params
+    stdc_downsample_mode: dw_conv   # STDC stage params
+
+
+context_module: aspp
```
Comment (ofrimasad) [inline src/super_gradients/recipes/arch_params/unet_default_arch_params.yaml:18]: this should look like the following:
```
context_module: aspp
    dilation_list: [2, 4, 6]
    in_out_ratio: 1.
    use_bias: False
    in_channels:  512  # currently it is manually inserted as lats width of backbone 
```
you just need to use `resolve_paramter` (see my comment in unet.py)

File: src/super_gradients/training/models/segmentation_models/unet/unet.py [hunk 14]
```diff
@@ -0,0 +1,231 @@
...
+            - aux_out_channels: List[int], Output channels, can be refers as num_classes, of auxiliary head per encoder
+                stage.
+        :param dropout: dropout probability of segmentation and auxiliary heads.
+        """
+        super().__init__(use_aux_heads=use_aux_heads)
+        self.num_classes = num_classes
+        # Init Backbone
+        backbone = UNetBackboneBase(**backbone_params)
+        # Init Context Module
+        context_module = nn.Identity() if context_module is None else build_context_module(
```
Comment (ofrimasad) [inline src/super_gradients/training/models/segmentation_models/unet/unet.py:86]: this piece of code is redundant. use `@resolve_parameter(context_module, ContextModuleFactory)` 
in_channels is known from backbone_params and can be defined in YAML

File: src/super_gradients/training/models/segmentation_models/unet/unet_decoder.py (hunk 4 above)
Comment (ofrimasad) [reply src/super_gradients/training/models/segmentation_models/unet/unet_decoder.py:82]: ok. very nice

**Review by ofrimasad — COMMENTED** [review]
looks very nice.
one last fix and it will be perfect

File: src/super_gradients/recipes/arch_params/unet_default_arch_params.yaml (hunk 13 above)
Comment (lkdci) [reply src/super_gradients/recipes/arch_params/unet_default_arch_params.yaml:18]: So you want in_channels to be manually hardcoded, and not dynamically adapted to the backbone?

File: src/super_gradients/training/models/segmentation_models/unet/unet.py (hunk 14 above)
Comment (lkdci) [reply src/super_gradients/training/models/segmentation_models/unet/unet.py]: Done [6bf7a2b](https://github.com/Deci-AI/super-gradients/pull/401/commits/6bf7a2bb76c41d8af67f4390f2da9bc9c0d59ddb)

**Comment by lkdci** [comment]
@ofrimasad , Addressed. Init the context module through a Yaml base factory instead of dynamically initiate w.r.t to the backbone.

**Review by ofrimasad — APPROVED** [review]
LGTM

**Review by ofrimasad — APPROVED** [review]
LGTM
